Separate clangimporter errors

adrian-prantl opened the pull request:
SwiftASTContext resets the state of the DiagnosticConsumer after each
import to make failed module imports non-fatal, however, the internal
state of the DiagnosticEngine in Swift's embedded Clang compiler does
not reset when encountering a fatal error. By storing Clang
diagnostics separately, these errors can be surfaces when an
expression fails to IRGen (which is usually when these fatal Clang
diagnostics turn into an unrecoverable problem).

rdar://108557254

adrian-prantl committed:
Simplify initializers (NFC)

adrian-prantl committed:
Simplify StoringDiagnosticConsumer::handleDiagnostic() (NFC)

adrian-prantl committed:
Unify all mechanisms for delaying diagnostics in SwiftASTContext (NFC)

adrian-prantl commented:
@swift-ci test

JDevlieghere commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp:
```diff
@@ -1969,30 +1971,21 @@ SwiftExpressionParser::Parse(DiagnosticManager &diagnostic_manager,
-            : "Please check the above error messages for possible root causes.");
+        eDiagnosticSeverityError,
+        "couldn't IRGen expression. Please enable the expression log by "
+        "running \"log enable lldb xpr\", then run the failing expression "
```
`s/xpr/expr/`

JDevlieghere commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp:
```diff
@@ -1663,7 +1663,9 @@ SwiftExpressionParser::Parse(DiagnosticManager &diagnostic_manager,
   unsigned buffer_id = UINT32_MAX;
   auto DiagnoseSwiftASTContextError = [&]() {
-    assert(m_swift_ast_ctx.HasErrors() && "error expected");
+    assert((m_swift_ast_ctx.HasErrors() ||
```
Would it be useful to have a method `HasAnyErrors` in `SwiftASTContext` which returns `HasErrors() || HasClangImporterErrors()` ?

adrian-prantl replied:
I thought about that, but I decided that it's better to be explicit about wanting to include ClangImporter errors at the call site.

JDevlieghere commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp:
```diff
@@ -1969,30 +1971,21 @@ SwiftExpressionParser::Parse(DiagnosticManager &diagnostic_manager,
+  // Clang diagnostic.
+  if (m_swift_ast_ctx.HasErrors() || m_swift_ast_ctx.HasClangImporterErrors()) {
+    diagnostic_manager.PutString(eDiagnosticSeverityRemark,
+                                 "couldn't IRGen expression.");
```
The fact that no error is provided is pretty interesting and probably worth mentioning to the user (if only to show that we couldn't do better). How about "couldn't IRGen expression: failed without error message" or something along those lines?

JDevlieghere replied:
Maybe this should go below where we encourage the user to enable logging.

adrian-prantl replied:
That's not what's happening here. The case where there is no error message is the one below where we encourage the logging.

JDevlieghere commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftUserExpression.cpp:
```diff
@@ -693,7 +693,7 @@ bool SwiftUserExpression::Parse(DiagnosticManager &diagnostic_manager,
 
   if (m_swift_ast_ctx->HasFatalErrors()) {
-    m_swift_ast_ctx->PrintDiagnostics(diagnostic_manager);
+    m_swift_ast_ctx->PrintDiagnostics(diagnostic_manager, false);
```
What's the meaning of the second argument? It would be helpful to add an inline comment `/*ArgName=*/`

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h:
```diff
@@ -423,19 +428,34 @@ class SwiftASTContext : public TypeSystemSwift {
+  bool HasErrors() const;
+  bool HasClangImporterErrors() const;
+
+  void RaiseDiagnostic(DiagnosticSeverity severity, llvm::StringRef message);
```
"Raise" sounds like you're increasing the severity level. How about `ReportDiagnostic` and `ReportFatalError`. That would be consistent with the `Report*` methods in `Debugger` for example.

adrian-prantl replied:
I'll go with AddDiagnostic for consistency.

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h:
```diff
@@ -423,19 +428,34 @@ class SwiftASTContext : public TypeSystemSwift {
+    return m_fatal_errors.Fail() || HasFatalErrors(m_ast_context_ap.get());
+  }
+
+  Status GetAllErrors() const;
```
Can this return an `llvm::Error` (that must be checked) instead?

adrian-prantl replied:
GetAllErrors() doesn't consume the errors, so I don't think that's needed.

JDevlieghere commented on lldb/source/Target/Target.cpp:
```diff
@@ -2474,7 +2474,7 @@ Target::GetScratchTypeSystemForLanguage(lldb::LanguageType language,
             else if (new_swift_ast_ctx->HasFatalErrors()) {
               DiagnosticManager diag_mgr;
-              new_swift_ast_ctx->PrintDiagnostics(diag_mgr);
+              new_swift_ast_ctx->PrintDiagnostics(diag_mgr, true);
```
Same question: what's the meaning of `true`?

adrian-prantl committed:
Replace all mutation of m_fatal_error with calls to RaiseFatalError() (NFC)

adrian-prantl commented:
@swift-ci test

adrian-prantl committed:
Unconditionally initialize StoringDiagnosticConsumer

adrian-prantl commented:
@swift-ci test

adrian-prantl committed:
Report errors in PerformCompileUnitImports to the user in addition to logging them.

adrian-prantl committed:
Introduce a "transactional" diagnostic consumer.

Swift compiler and ClangImporter and LLDB operations can produce
diagnostics, which are stored in StoringDiagnosticConsumer. Operations
such as expression evaluation of module imports may produce multiple
non-fatal error diagnostics.  Prior to this patch, the expression
parser would clear the diagnostic consumer before running. This had
the unwanted side-effect of sometimes clearing out unreported earlier
diagnostics.

This patch introduces SwiftASTContext::forkDiagnosticConsumer() which
returns an RAII object that behaves like a diagnostic consumer, can
tell whether diagnostics where produced during its lifetime, and
resets the StoringDiagnosticConsumer to its previous state upon
destruction. This way

- unreported diagnostics don't get cleared by the expression parser
- the expression parser has precise access to only the diagnostics it produced
- while still being able to report persistent errors
  (such as clang import failures) that happened before the expression
  parser started.

I'm not happy by the amount of code needed to make this work, or the
specific implementation, but at least the interface provided to the
outside is nice and clean.

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
Introduce a "transactional" diagnostic consumer.

Swift compiler and ClangImporter and LLDB operations can produce
diagnostics, which are stored in StoringDiagnosticConsumer. Operations
such as expression evaluation of module imports may produce multiple
non-fatal error diagnostics.  Prior to this patch, the expression
parser would clear the diagnostic consumer before running. This had
the unwanted side-effect of sometimes clearing out unreported earlier
diagnostics.

This patch introduces SwiftASTContext::forkDiagnosticConsumer() which
returns an RAII object that behaves like a diagnostic consumer, can
tell whether diagnostics where produced during its lifetime, and
resets the StoringDiagnosticConsumer to its previous state upon
destruction. This way

- unreported diagnostics don't get cleared by the expression parser
- the expression parser has precise access to only the diagnostics it produced
- while still being able to report persistent errors
  (such as clang import failures) that happened before the expression
  parser started.

I'm not happy by the amount of code needed to make this work, or the
specific implementation, but at least the interface provided to the
outside is nice and clean.

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test windows

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -910,16 +911,469 @@ static std::string GetClangModulesCacheProperty() {
+      if (severity == eDiagnosticSeverityError && diagnostic.line == 0) {
+        diagnostic_manager.AddDiagnostic(diagnostic.description.c_str(),
+                                         severity, origin);
+        added_one_diagnostic = true;return;
```
missing newline

adrian-prantl commented:
@swift-ci test

bulbazord commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp:
```diff
@@ -1969,30 +1972,22 @@ SwiftExpressionParser::Parse(DiagnosticManager &diagnostic_manager,
+  if (expr_diagnostics->HasErrors() ||
+      m_swift_ast_ctx.HasClangImporterErrors()) {
+    diagnostic_manager.PutString(eDiagnosticSeverityRemark,
+                                 "couldn't IRGen expression");
```
Are these messages meant to be seen by end users? If so, IRGen as a term is pretty specific to clang, swift, and llvm... maybe we can say something a little more user friendly like:

Failed to compile expression: IR generation failure

adrian-prantl replied:
That's a valid concern! Note that this is just one remark in a longer error. For example, the test in `lang/swift/clangimporter/clang_errorhandling` produces:

```
error: expression failed to parse:
couldn't IRGen expression
error: /Volumes/Data/swift-5.9/llvm-project/lldb/test/API/lang/swift/clangimporter/clang_errorhandling/bridging-header.h:2:1: error: unknown type name 'i'
i am a syntax error
^

error: /Volumes/Data/swift-5.9/llvm-project/lldb/test/API/lang/swift/clangimporter/clang_errorhandling/bridging-header.h:2:5: error: expected ';' after top level declarator
i am a syntax error
    ^

error: failed to import bridging header '/Volumes/Data/swift-5.9/llvm-project/lldb/test/API/lang/swift/clangimporter/clang_errorhandling/bridging-header.h'
```
as the full error message.

adrian-prantl replied:
The IRGen remark is really mostly there for LLDB developers to understand at what point the expression failed to help with troubleshooting.

adrian-prantl replied:
With this in mind — any suggestions for how to improve the error message?

bulbazord commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp:
```diff
@@ -1969,30 +1972,22 @@ SwiftExpressionParser::Parse(DiagnosticManager &diagnostic_manager,
-              "bugreport with the log output."
-            : "Please check the above error messages for possible root causes.");
+        eDiagnosticSeverityError,
+        "couldn't IRGen expression. Please enable the expression log by "
+        "running \"log enable lldb expr\", then run the failing expression "
+        "again, and file a bug report with the log output.");
```
Same here as above

adrian-prantl replied:
The idea is that nobody should ever see this, and if you see this error message there's something that LLDB developers need to investigate and the error message is meant for them, not for the user.

bulbazord commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -910,16 +911,470 @@ static std::string GetClangModulesCacheProperty() {
+  ///        used to change only the bold attribute, and keep colors
+  ///        untouched.
+  /// \param Bold bold/brighter text, default false
+  /// \param BG if true change the background,
```
I think the parameter names in the documentation are incorrect?

bulbazord commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -910,16 +911,470 @@ static std::string GetClangModulesCacheProperty() {
+    std::string formatted_text;
+    if (!line_col.first) {
+      formatted_text = text.str();
+      use_fixits = false;
```
Nit: it's already false from a few lines up.

bulbazord reviewed (commented):
Haven't looked at all of SwiftASTContext yet but the other stuff looks ok I think.

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h:
```diff
@@ -423,20 +428,65 @@ class SwiftASTContext : public TypeSystemSwift {
+    /// object.
+    Status GetAllErrors() const;
+  };
+  std::unique_ptr<TransactionDiagnostics> forkDiagnosticConsumer();
```
The `fork` part made me wonder if this was doing anything special to warrant that name, but it doesn't seem like it. Why not `CreateDiagnosticConsumer`?

adrian-prantl replied:
Because it's an overlay to the main diagnostic consumer. All the diagnostics are still handled by the parent consumer, but it deletes certain kinds of diagnostics after destruction.

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h:
```diff
@@ -423,20 +428,65 @@ class SwiftASTContext : public TypeSystemSwift {
+
+  /// A lightweigt abstraction that stits on top of a diagnostic
+  /// consumer that can be used capture diagnostics for one transation
+  /// and restore the state of the consumer afterwards.
```
This sounds like a RAII object. Might be worth using that term here? I had to read the comment to understand what this class is doing. I think `ScopedDiagnostics` might be more obvious for someone who's familiar with other Scoped* classes in LLVM/LLDB.

adrian-prantl replied:
Correct! I'll try ScopedDiagnostics and see how that looks.

JDevlieghere commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp:
```diff
@@ -1373,16 +1372,16 @@ static llvm::Expected<ParsedExpression> ParseAndImport(
 
-  if (swift_ast_context.HasErrors())
+  if (expr_diagnostics.HasErrors())
     return make_error<SwiftASTContextError>();
```
Seems unfortunate to not percolate anything up here?

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -910,16 +911,470 @@ static std::string GetClangModulesCacheProperty() {
+
+namespace lldb_private {
+
+class ANSIColorStringStream : public llvm::raw_string_ostream {
```
Nit: seems like this really should be its own header.

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -910,16 +911,470 @@ static std::string GetClangModulesCacheProperty() {
+  bool m_colorize;
+};
+
+class StoringDiagnosticConsumer : public swift::DiagnosticConsumer {
```
Same comment as above. SwiftASTContext is already huge, why not move this into its own file?

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -910,16 +911,470 @@ static std::string GetClangModulesCacheProperty() {
+    // numbers around by adjusting the source location, we must do it
+    // manually. We also want to use the same error formatting as LLVM
+    // and Clang, so we must muck with the string.
+
```
Spurious newline?

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -910,16 +911,470 @@ static std::string GetClangModulesCacheProperty() {
+         consumer.m_raw_clang_diagnostics.size() > m_cursor.clang;
+}
+
+Status SwiftASTContext::TransactionDiagnostics::GetAllErrors() const {
```
Why not an `Error` that must be checked?

adrian-prantl replied:
Done.

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -910,16 +911,470 @@ static std::string GetClangModulesCacheProperty() {
+  auto &lldb_diags = consumer.m_diagnostics;
+  auto &swift_diags = consumer.m_raw_swift_diagnostics;
+
+  // Intentionally don't reset clang diagnostics.
```
Why?

adrian-prantl replied:
I added a better comment.

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -910,16 +911,470 @@ static std::string GetClangModulesCacheProperty() {
+
+SwiftASTContext::TransactionDiagnostics::TransactionDiagnostics(
+    swift::DiagnosticConsumer &consumer)
+    : m_consumer(consumer),
```
You could make `m_consumer` a `StoringDiagnosticConsumer` and then you'd only have to do the cast oncee.

adrian-prantl replied:
I'm trying to not introduce any Swift compiler dependencies into SwiftASTContext.h

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -910,16 +911,470 @@ static std::string GetClangModulesCacheProperty() {
+
+Status SwiftASTContext::TransactionDiagnostics::GetAllErrors() const {
+  // Retrieve the error message from the DiagnosticConsumer.
+  DiagnosticManager diagnostic_manager;
```
I'm confused: isn't this operating on a local copy?

adrian-prantl replied:
DiagnosticManager != StoringDiagnosticConsumer

This is printing the contents of the diagnosic consumer into a fresh lldb::DiagnosticManager to then dump it into Status.

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -910,16 +911,470 @@ static std::string GetClangModulesCacheProperty() {
+    if (consumer.m_diagnostics[i]->GetSeverity() == eDiagnosticSeverityError)
+      num_errors++;
+
+  return num_errors > 0 ||
```
You don't need to count the number of errors if the last two conditions are true. You could do an early return:

```
 if (consumer.m_num_swift_errors > m_cursor.m_num_swift_errors ||
          consumer.m_raw_clang_diagnostics.size() > m_cursor.clang)
  return true;
```

and compute `num_errors` after.

adrian-prantl committed:
Remove redundant initialization

adrian-prantl commented:
@swift-ci test

adrian-prantl committed:
Surface LLVM verification errors.

adrian-prantl committed:
Convert GetAllErrors() to return an llvm::Error (NFC)

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test windows

JDevlieghere reviewed (approved):
I didn't re-review everything but I see my concerns were addressed to LGTM.

adrian-prantl committed:
Separate Swift and Clang errors in StoringDiagnosticConsumer

SwiftASTContext resets the state of the DiagnosticConsumer after each
import to make failed module imports non-fatal, however, the internal
state of the DiagnosticEngine in Swift's embedded Clang compiler does
not reset when encountering a fatal error. By storing Clang
diagnostics separately, these errors can be surfaces when an
expression fails to IRGen (which is usually when these fatal Clang
diagnostics turn into an unrecoverable problem).

rdar://108557254

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test windows
